### ISSUE TITLE
spawn rate suggestion

### ISSUE THREAD

**Issue by jaxjaxk** [Zet0rz/nzombies#259]
i believe it would be right to increase the spawn rate of zombies based off of how many players are "ready" ( ie not spectating). the problem is people camp a whole lot because the zombie spawn rate is the same and with more players it is easy to camp together and never be overwhelmed even in later rounds. it also means that rounds take very long because more players = more zombies spawned.

i proposed something like this

new_spawn_rate = default_spawn_rate \* sqrt(#plays)

where default_spawn_rate is what we have it set to now and #plys is the number of players that are "ready"

i believe for now using sqrt would be best as it does not change the rate by much and this is something that would change the difficulty a lot even if changed by a little so baby steps with the spawn rate.

so tell me what you think.

**Comment by lolleko** [comment]
There is no such thing as a spawn rate, Zombies spawn as soon there is a non occupied spawnpoint nearby.

**Comment by jaxjaxk** [comment]
@lolleko 
if that were true then this `self:GetSpawner():SetNextSpawn(CurTime() + self:GetSpawner():GetDelay())` would not exist.

not only that but if their was no delay then more spawns would be activated and spawn in zombies quicker.

**Comment by lolleko** [comment]
> if that were true then this self:GetSpawner():SetNextSpawn(CurTime() + self:GetSpawner():GetDelay()) would not exist.

I wrote that system 😄 , the delay is just there to prevent one spawner from becoming to dominant. The delay in the main system is 0.25 seconds by default so if there are 4 valid spawn points 4 zombies could spawn in one second. Which is fine since most of the time you won't have 4 valid spawns in one second. You could lower that delay but it wouldn't make change the "spawnrate". Try to place more spawns, if you have camping issues. I mean if you only have 4 spawns 4 players are enough to camp those 4 spawns.

> not only that but if their was no delay then more spawns would be activated and spawn in zombies quicker.

Spawns being a activated relies on many factors not just the delay, as i explained above the delay is really small, so changing it wouldn't make much difference.

tl;dr Zombies spawn (almost) as fast as possible.

**Comment by jaxjaxk** [comment]
@lolleko 
mate i have 16 people on and roughly 30-50 spawns on a map. it is not enough.

i have reviewed the code and honestly im amazed and disappointed at the same time.

``` lua
    local zombiesToSpawn = self.iZombiesToSpawn / 2
    local totalDistributed = 0
    for k, vspawn in pairs(self.tValidSpawns) do
        if k < #self.tValidSpawns then
            vspawn:SetZombiesToSpawn(math.ceil(zombiesToSpawn))
            totalDistributed = totalDistributed + math.ceil(zombiesToSpawn)
            zombiesToSpawn = math.floor(zombiesToSpawn / 2)
        else
            -- add the remaining zombies to the last spawn in list
            vspawn:SetZombiesToSpawn(self.iZombiesToSpawn - totalDistributed)
        end
    end
```

with 100 zombies to spawn the max amount of spawners possible is only 7. what is worse is when you get to 200 the max possible spanners is now 8. naturally the pattern continues to 400 only letting up to 9 spawners active.

**Comment by Zet0rz** [comment]
I have noticed that in multiplayer, spawning of zombies can be very slow and highly favor certain spawnpoints over others. I have realized that we shouldn't calculate weight based on the average distance of spawnpoint to players, but rather the shortest distance to any player. That combined with zombie distribution being calculated from their weight rather than halved every time might improve this.

**Comment by minin43** [comment]
Can we just make the spawns more random? I don't want to be able to sit next to one spawn point and be guaranteed they'll always spawn there because it's the shortest distance away

**Comment by jaxjaxk** [comment]
@Zet0rz 
so far i have revised it to distribute the zombies evenly between valid spawns and improved the number of valid spawns by using "spawn:GetSpawnWeight()/(math.sqrt(#plys)) <= average" so that more spawns can be added with more players. while it works i do not have much data on how well it works yet but will continue testing.

**Comment by jaxjaxk** [comment]
@minin43 i like this idea a lot more however it kind of makes the zombies more spread out depending on the map.

if i were to be insane i would suggest we create several methods of spawning and create an option in the map options on what method to use. now THAT would we epic but hard as hell to pull off.

**Comment by Zet0rz** [comment]
The way I will probably do it as I suggested should result in a similar spawn distribution as what we have right now, except individually for each player. So a player far away from the group will activate all the spawners there as much as the group will back where they are. Distributing zombies based on the spawners weight should mean that _all_ spawners within range would be active, those further away spawning less (but still some).

> spawn:GetSpawnWeight()/(math.sqrt(#plys)) <= average

That could work for a bit of fine-tuning, but I think shortest-distance weight calculations will fix multiplayer weight problems (average is still going to be average for all players)

> Can we just make the spawns more random? I don't want to be able to sit next to one spawn point and be guaranteed they'll always spawn there because it's the shortest distance away

That's not how they are distributed. They will spawn on more than 1 dominant spawnpoint given there are others close enough. But it is not as much of a random chance as it is a distribution of zombies to spawn.

**Comment by lolleko** [comment]
I thought this issue was about how fast zombies spawn, not how many. You are right the amount of active zombie spawns is `log2(zombiesToSpawn)` which worked fine on the small map i tested with back then.
But @jaxjaxk  is right for many players and larger maps that growth is way to small. I will fix the distribution to be mor even so mor spawners get activated.

**Comment by Zet0rz** [comment]
Distribute the zombies to spawn by weight (could be done instead of sorting the table), e.g. `maxzombies * (vspawn:GetWeight()/totalweight)`, add some rounding and ensure max zombie count is achieved.

Set weight based on minimum distance to any player rather than average of all players. Will make it so that a single player in one side of the map will still get zombies spawning even if there are 10 players in the opposite side (which would actually as it is now set player 1's spawnpoints' weights even lower than average, so there'd be NO zombies by him)

**Comment by Zet0rz** [comment]
Spawn distribution changed in 2341c9d712ad612d2c18e16821a71bc701b71115. It is now handled based on shortest distance, and this also decides how many zombies it spawns (based on a percentage of the total weight from all active spawnpoints)

It isn't finalized, so give it some testing. It could go wrong or still not work properly. And it will likely not have a big effect if all players are grouped together (its main point was to fix when players are seperated).
